apps/prairielearn/src/lib/editors.ts

```diff
@@ -193,41 +199,62 @@ export abstract class Editor {
           env: gitEnv,
         });
 
-        job.info('Fetch from remote git repository');
-        await job.exec('git', ['fetch'], {
-          cwd: this.course.path,
-          env: gitEnv,
-        });
-
         await cleanAndResetRepository(this.course, gitEnv, job);
 
+        const writeAndCommitChanges = async () => {
+          job.data.saveAttempted = true;
+
+          if (!(await this.shouldEdit())) return;
+
+          job.info('Write changes to disk');
+          const { pathsToAdd, commitMessage } = await this.write();
+
+          job.info('Commit changes');
+          await job.exec('git', ['add', ...pathsToAdd], {
+            cwd: this.course.path,
+            env: gitEnv,
+          });
+          await job.exec(
+            'git',
+            [
+              '-c',
+              `user.name="${this.user.name}"`,
+              '-c',
+              `user.email="${this.user.email || this.user.uid}"`,
+              'commit',
+              '-m',
+              commitMessage,
+            ],
+            {
+              cwd: this.course.path,
+              env: gitEnv,
+            },
+          );
+        };
+
         try {
-          if (await this.shouldEdit()) {
-            job.data.saveAttempted = true;
-            job.info('Write changes to disk');
-            const { pathsToAdd, commitMessage } = await this.write();
+          await writeAndCommitChanges();
+
+          try {
+            job.info('Push changes to remote git repository');
+            await job.exec('git', ['push'], {
+              cwd: this.course.path,
+              env: gitEnv,
+            });
+            job.data.saveSucceeded = true;
+          } catch (err) {
+            job.info('Failed to push changes to remote git repository');
+            job.info('Pulling changes from remote git repository and trying again');
 
-            job.info('Commit changes');
-            await job.exec('git', ['add', ...pathsToAdd], {
+            job.info('Fetch from remote git repository');
+            await job.exec('git', ['fetch'], {
               cwd: this.course.path,
               env: gitEnv,
             });
-            await job.exec(
-              'git',
-              [
-                '-c',
-                `user.name="${this.user.name}"`,
-                '-c',
-                `user.email="${this.user.email || this.user.uid}"`,
-                'commit',
-                '-m',
-                commitMessage,
-              ],
-              {
-                cwd: this.course.path,
-                env: gitEnv,
-              },
-            );
+
+            await cleanAndResetRepository(this.course, gitEnv, job);
```

Comment:
Just to make sure I'm understanding, we are relying on the reset here to throw away our first commit that failed to push?

Reply:
Correct. I added a comment explaining this in 40f957a.

Reply:
Perfect, thanks.